source/pape-manual-pages/rblsmtpd.1
.TH rblsmtpd 1
.SH NAME
rblsmtpd \- blocks mail from RBL-listed sites. It works with any SMTP server that can run under
.BR tcpserver (1)
.SH SYNOPSIS
.B rblsmtpd
.I opts
.I prog
.SH DESCRIPTION
.I opts
is a series of getopt-style options.
.I prog
consists of one or more arguments.

Normally
.B rblsmtpd
runs
.IR prog .
.I prog
is expected to carry out an SMTP conversation to receive incoming mail messages. 

However,
.B rblsmtpd
does not invoke
.I prog
if it is told to block mail from this client. Instead it carries out its own limited SMTP conversation, temporarily rejecting all attempts to send a message. Meanwhile it prints one line on descriptor 2 to log its activity.

.B rblsmtpd
drops the limited SMTP conversation after 60 seconds, even if the client has not quit by then.
.SH OPTIONS
.TP
.B \-t \fIn
Change the timeout to
.I n
seconds.
.P
.B Blocked clients
.P
If the $RBLSMTPD environment variable is set and is nonempty,
.B rblsmtpd
blocks mail. It uses $RBLSMTPD as an error message for the client. Normally
.B rblsmtpd
runs under
.BR tcpserver (1);
you can use
.BR tcprules (1)
to set $RBLSMTPD for selected clients. 

If $RBLSMTPD is set and is empty,
.B rblsmtpd
does not block mail. 

If $RBLSMTPD is not set,
.B rblsmtpd
looks up $TCPREMOTEIP in the RBL, and blocks mail if $TCPREMOTEIP is listed.
.B tcpserver
sets up $TCPREMOTEIP as the IP address of the remote host.
.TP
.B \-r \fIbase
Use
.I base
as an RBL source. An IP address
.I a.b.c.d
is listed by that source if
.I d.c.b.a.base
has a TXT record.
.B rblsmtpd
uses the contents of the TXT record as an error message for the client.
.TP
.B \-a \fIbase
Use
.I base
as an anti-RBL source. An IP address
.I a.b.c.d
is anti-listed by that source if
.I d.c.b.a.base
has an A record. In this case
.B rblsmtpd
does not block mail.
.P
You may supply any number of
.B \-r
and
.B \-a
options.
.B rblsmtpd
tries each source in turn until it finds one that lists or anti-lists $TCPREMOTEIP. It also tries an RBL source of rbl.maps.vix.com if you do not supply any
.B -r
options. See http://maps.vix.com/rbl/ for more information about rbl.maps.vix.com. 

If you want to run your own RBL source or anti-RBL source for
.BR rblsmtpd ,
you can use
.B rbldns
from the DNScache (djbdns) package.
.P
.B Temporary errors
.P
Normally, if $RBLSMTPD is set,
.B rblsmtpd
uses a 451 error code in its limited SMTP conversation. This tells legitimate clients to try again later. It gives innocent relay operators a chance to see the problem, prohibit relaying, get off the RBL, and get the mail delivered. 

However, if $RBLSMTPD begins with a hyphen,
.B rblsmtpd
removes the hyphen and uses a 553 error code. This tells legitimate clients to bounce the message immediately. 

There are several error-handling options for RBL lookups:
.TP
.B \-B
(Default.) Use a 451 error code for IP addresses listed in the RBL.
.TP
.B \-b
Use a 553 error code for IP addresses listed in the RBL.
.TP
.B \-C
(Default.) Handle RBL lookups in a ``fail-open'' mode. If an RBL lookup fails temporarily, assume that the address is not listed; if an anti-RBL lookup fails temporarily, assume that the address is anti-listed. Unfortunately, a knowledgeable attacker can force an RBL lookup or an anti-RBL lookup to fail temporarily, so that his mail is not blocked.
.TP
.B \-c
Handle RBL lookups in a ``fail-closed'' mode. If an RBL lookup fails temporarily, assume that the address is listed (but use a 451 error code even with -b). If an anti-RBL lookup fails temporarily, assume that the address is not anti-listed (but use a 451 error code even if a subsequent RBL lookup succeeds with -b). Unfortunately, this sometimes delays legitimate mail.
.SH SEE ALSO
tcpserver(1),
tcprules(1),
tcprulescheck(1),
fixcrio(1),
recordio(1),
rblsmtpd(1),
tcpclient(1),
who@(1),
date@(1),
finger@(1),
http@(1),
tcpcat(1),
mconnect(1),
tcp-environ(5)

http://cr.yp.to/ucspi-tcp.html


source/rblsmtpd.c
#include "byte.h"
#include "str.h"
#include "scan.h"
#include "fmt.h"
#include "env.h"
#include "exit.h"
#include "sig.h"
#include "buffer.h"
#include "readwrite.h"
#include "sgetopt.h"
#include "strerr.h"
#include "stralloc.h"
#include "commands.h"
#include "pathexec.h"
#include "dns.h"

#define FATAL "rblsmtpd: fatal: "

void nomem(void)
{
  strerr_die2x(111,FATAL,"out of memory");
}
void usage(void)
{
  strerr_die1x(100,"rblsmtpd: usage: rblsmtpd [ -b ] [ -R ] [ -t timeout ] [ -r base ] [ -a base ] smtpd [ arg ... ]");
}

char *ip_env;
static stralloc ip_reverse;

void ip_init(void)
{
  unsigned int i;
  unsigned int j;

  ip_env = env_get("TCPREMOTEIP");
  if (!ip_env) ip_env = "";

  if (!stralloc_copys(&ip_reverse,"")) nomem();

  i = str_len(ip_env);
  while (i) {
    for (j = i;j > 0;--j) if (ip_env[j - 1] == '.') break;
    if (!stralloc_catb(&ip_reverse,ip_env + j,i - j)) nomem();
    if (!stralloc_cats(&ip_reverse,".")) nomem();
    if (!j) break;
    i = j - 1;
  }
}

unsigned long timeout = 60;
int flagrblbounce = 0;
int flagfailclosed = 0;
int flagmustnotbounce = 0;

int decision = 0; /* 0 undecided, 1 accept, 2 reject, 3 bounce */
static stralloc text; /* defined if decision is 2 or 3 */

static stralloc tmp;

void rbl(char *base)
{
  if (decision) return;
  if (!stralloc_copy(&tmp,&ip_reverse)) nomem();
  if (!stralloc_cats(&tmp,base)) nomem();
  if (dns_txt(&text,&tmp) == -1) {
    flagmustnotbounce = 1;
    if (flagfailclosed) {
      if (!stralloc_copys(&text,"temporary RBL lookup error")) nomem();
      decision = 2;
    }
    return;
  }
  if (text.len) {
    if (flagrblbounce)
      decision = 3;
    else
      decision = 2;
  }
}

void antirbl(char *base)
{
  if (decision) return;
  if (!stralloc_copy(&tmp,&ip_reverse)) nomem();
  if (!stralloc_cats(&tmp,base)) nomem();
  if (dns_ip4(&text,&tmp) == -1) {
    flagmustnotbounce = 1;
    if (!flagfailclosed)
      decision = 1;
    return;
  }
  if (text.len)
    decision = 1;
}

char strnum[FMT_ULONG];
static stralloc message;

char inspace[64]; buffer in = BUFFER_INIT(read,0,inspace,sizeof inspace);
char outspace[1]; buffer out = BUFFER_INIT(write,1,outspace,sizeof outspace);

void reject() { buffer_putflush(&out,message.s,message.len); }
void accept() { buffer_putsflush(&out,"250 rblsmtpd.local\r\n"); }
void greet()  { buffer_putsflush(&out,"220 rblsmtpd.local\r\n"); }
void quit()   { buffer_putsflush(&out,"221 rblsmtpd.local\r\n"); _exit(0); }
void drop()   { _exit(0); }

struct commands smtpcommands[] = {
  { "quit", quit, 0 }
, { "helo", accept, 0 }
, { "ehlo", accept, 0 }
, { "mail", accept, 0 }
, { "rset", accept, 0 }
, { "noop", accept, 0 }
, { 0, reject, 0 }
} ;

void rblsmtpd(void)
{
  int i;

  if (flagmustnotbounce || (decision == 2)) {
    if (!stralloc_copys(&message,"451 ")) nomem();
  }
  else
    if (!stralloc_copys(&message,"553 ")) nomem();

  if (text.len > 200) text.len = 200;
  if (!stralloc_cat(&message,&text)) nomem();
  for (i = 0;i < message.len;++i)
    if ((message.s[i] < 32) || (message.s[i] > 126))
      message.s[i] = '?';
  
  buffer_puts(buffer_2,"rblsmtpd: ");
  buffer_puts(buffer_2,ip_env);
  buffer_puts(buffer_2," pid ");
  buffer_put(buffer_2,strnum,fmt_ulong(strnum,getpid()));
  buffer_puts(buffer_2,": ");
  buffer_put(buffer_2,message.s,message.len);
  buffer_puts(buffer_2,"\n");
  buffer_flush(buffer_2);

  if (!stralloc_cats(&message,"\r\n")) nomem();

  if (!timeout)
    reject();
  else {
    sig_catch(sig_alarm,drop);
    alarm(timeout);
    greet();
    commands(&in,smtpcommands);
  }
  _exit(0);
}

int main(int argc,char **argv,char **envp)
{
  int flagwantdefaultrbl = 1;
  char *x;
  int opt;

  ip_init();

  x = env_get("RBLSMTPD");
  if (x) {
    if (!*x)
      decision = 1;
    else if (*x == '-') {
      if (!stralloc_copys(&text,x + 1)) nomem();
      decision = 3;
    }
    else {
      if (!stralloc_copys(&text,x)) nomem();
      decision = 2;
    }
  }

  while ((opt = getopt(argc,argv,"bBcCt:r:a:")) != opteof)
    switch(opt) {
      case 'b': flagrblbounce = 1; break;
      case 'B': flagrblbounce = 0; break;
      case 'c': flagfailclosed = 1; break;
      case 'C': flagfailclosed = 0; break;
      case 't': scan_ulong(optarg,&timeout); break;
      case 'r': rbl(optarg); flagwantdefaultrbl = 0; break;
      case 'a': antirbl(optarg); break;
      default: usage();
    }

  argv += optind;
  if (!*argv) usage();

  if (flagwantdefaultrbl) rbl("rbl.maps.vix.com");
  if (decision >= 2) rblsmtpd();

  pathexec_run(*argv,argv,envp);
  strerr_die4sys(111,FATAL,"unable to run ",*argv,": ");
}
